### PR TITLE
Support per-motor throttle-based harmonic notch

### PR DESCRIPTION
**Pull request by andyp1per**
To enable set the mult-source bit in INS_HNTCx_OPTS and use the throttle notch

### CODE REVIEW

**Review by amilcarlucas — APPROVED** [review]

File: libraries/AP_Motors/AP_MotorsMatrix.cpp
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: I think you will find `_thrust_rpyt_out` is wrong when motors are not in throttle unlimited. You should be able to use `_actuator` (which also obeys slew limiting) there is a `thr_lin.actuator_to_thrust` method to remove the expo to get back to the original value. Because that is a MotorsMulticopter level variable you should be able to move this up and have it work for tricopter ect too.

Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: Also you will need a `abs` too in that case, the 6dof backed can output negative values.

Comment (lthall) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: Where do you think it is changed when throttle limited?

Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: Its simply that `_thrust_rpyt_out` is not output to the motors but `output_armed_stabilizing` is still run. So `_thrust_rpyt_out` could have some none zero stuff in it. It depends how good a job were doing keeping all the controllers zeroed (copter is probably better than plane).

Independent of that if you have some motor slew limits set then `_thrust_rpyt_out` by-passes the that in throttle unlimited. 

It would also be nice if the filters tacked up with spool up. Of course the lower harmonics would have zero attenuation thanks to the recent re-work, but the higher harmonics would be active. This is what would happen with RPM feedback, I presume the goal is to match that as closely as possible.

Comment (lthall) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: I agree with you that there are parts of this that slew rate factors that are not included however the complexity of removing slew rate and linearisation doesn't justify the gains in my opinion.

**Review by lthall — APPROVED** [review]
This all looks right to me.

**Comment by andyp1per** [comment]
Flight tested on a quad with ESC telemetry. First notch uses ESC telemetry, second notch uses throttle - both multi-source, outcome seems really pretty good:

image

**Comment by andyp1per** [comment]
I've added this to DevCall so that @IamPete1 and @lthall can duke it out.

File: libraries/AP_Motors/AP_MotorsMatrix.cpp
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: There are several benefits to using `_actuator`, I do think its worth doing. These are the ones I can think of at the moment:

- Slew limits, spool up and down, E-Stop and interlock all work correctly. 
- Works for all `AP_MotorsMulticopter` derived classes, not just `AP_MotorsMatrix`. Tricopter, coax, single, tailsitter.
- Works in forward flight for Quadplanes using the `output_motor_mask` method.

**Review by IamPete1 — CHANGES_REQUESTED** [review]
I really think this should be useing the `_actuator` value and removing the thrust linearisation with the `actuator_to_thrust` method.

See discussion: https://github.com/ArduPilot/ardupilot/pull/26674#discussion_r1548786199

File: libraries/AP_Motors/AP_MotorsMatrix.cpp
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:414]: The log from @andyp1per highlights the first point. ESC driven in green, throttle driven in red. The PWM output in blue. 

image

We see the throttle based in red go to zero while the vehicle is in ground idle. Where as the PWM output derived from the `_actuator` value tracks up and down with the true RPM. `_actuator` includes spin arm/min/max so can track in those cases. Although we might have to have a think about best remove the linearisation outside the spin range.

**Comment by IamPete1** [comment]
I have done the change over to use `_actuator` here: https://github.com/IamPete1/ardupilot/tree/actuator_throttle_notch

This is the key commit: https://github.com/IamPete1/ardupilot/commit/c70db8608845bd4b07b158e578529a48c3b4bc1b

**Comment by tridge** [comment]
@andyp1per nice work! I'd love to see some test logs

**Review by tridge — CHANGES_REQUESTED** [review]
i think we need a pre-arm check for the ratio of bw to freq for multi-source
we could have a notch filter option to disable the check

**Comment by tridge** [comment]
@andyp1per @lthall maybe we should pre-arm warn if you have multi-source and bw > 0.75*freq/nsources ? we could have a INS_HNTCH_OPTS bit to disable this check

**Comment by andyp1per** [comment]
> @andyp1per @lthall maybe we should pre-arm warn if you have multi-source and bw > 0.75*freq/nsources ? we could have a INS_HNTCH_OPTS bit to disable this check

We don't currently do this for ESC notches and I am currently against it. Please can we move the discussion of that to a separate PR so that we can have a good discussion and not hold this PR up?

**Comment by tridge** [comment]
> We don't currently do this for ESC notches and I am currently against it. Please can we move the discussion of that to a separate PR so that we can have a good discussion and not hold this PR up?

this PR (which I love btw!) will make this situation a lot more common. We need to fix it before we merge this PR. If we defer this test we'll never put it in and we'll get crashes

File: libraries/AP_Motors/AP_MotorsMatrix.cpp
Comment (andyp1per) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp]: Ok @IamPete1 's change checked by @lthall and incorporated

**Comment by andyp1per** [comment]
> i think we need a pre-arm check for the ratio of bw to freq for multi-source we could have a notch filter option to disable the check

This has been split out into https://github.com/ArduPilot/ardupilot/pull/27068

**Review by IamPete1 — APPROVED** [review]
LGTM

**Comment by tridge** [comment]
@andyp1per I put some logs here for a hexa that has ESC telemetry:
https://drive.google.com/drive/folders/1fveumuzs7zw7zCqw7jKLHqiaXxx3k5nC?usp=drive_link
I used a small lua script to allow switching between the ESC notch and the per-motor throttle notch:
https://github.com/ArduPilot/ardupilot/pull/27163
the match is OK for some motors, less good for others. Here is motor3 for example:
image
and motor6 is good too:
image
motor5 is less good:
image
this is all from 1.BIN, which was the last flight we did (microSD swap before this flight)
we couldn't try the octo as it has an orangeplus on it, so no bdshot for 8 motors

**Comment by andyp1per** [comment]
I had a look at the BobaHex logs using this. Unfortunately the ref was slightly wrong leading to about a 15Hz separation between the actual motor frequency and the frequency being used by the throttle notch. This has the effect of effectively doubling the bandwidth being used leading to better noise attenuation than would have otherwise be attained using the ESC notch only with a relatively narrow bandwidth (Q=6.2). I'm not sure this proves much other than flying with wider notches is ok. My goto notch width is Q=5 which is not far away so really need an octo test with narrower notch to see the differences.

image

**Review by lthall — APPROVED** [review]
This really should go in.

**Comment by andyp1per** [comment]
Thanks @tridge !

**Pull request merged**
